File: apps/cli/src/commands/analyze.rs
use crate::cli::Cli;
use crate::commands::{build, target};
use crate::error::{AppError, AppResult};
use crate::output::{self, Meta};
use serde::Serialize;
use std::process::Command;

#[derive(Debug, Serialize)]
struct AnalyzeData {
    project_root: String,
    status: String,
    build_status: String,
    test_status: String,
    diagnostics: Vec<build::Diagnostic>,
    findings: Vec<AnalyzeFinding>,
    test_stdout: String,
    test_stderr: String,
}

#[derive(Debug, Serialize)]
struct AnalyzeFinding {
    severity: String,
    source: String,
    message: String,
    file: Option<String>,
    line: Option<u64>,
    column: Option<u64>,
}

pub fn run(cli: &Cli) -> AppResult<()> {
    let data = analyze_data(cli)?;
    if cli.json {
        let mut meta = Meta::new("analyze");
        meta.project_root = Some(data.project_root.clone());
        output::print_json(data, meta)
    } else if data.status == "success" {
        println!("Analysis passed: {}", data.project_root);
        Ok(())
    } else {
        print_human_findings(&data);
        Err(AppError::user(
            "analysis_failed",
            "ConSol analysis found issues.",
            Some(format!("{} finding(s) reported.", data.findings.len())),
        ))
    }
}

fn analyze_data(cli: &Cli) -> AppResult<AnalyzeData> {
    let resolved = target::resolve(cli, None)?;
    let build = build::build_data(cli, None)?;
    let test = forge_test(&resolved.project_root)?;
    let mut findings = build
        .diagnostics
        .iter()
        .map(finding_from_diagnostic)
        .collect::<Vec<_>>();
    if test.status != "success" {
        findings.push(AnalyzeFinding {
            severity: "error".to_string(),
            source: "forge test".to_string(),
            message: "Foundry tests failed.".to_string(),
            file: None,
            line: None,
            column: None,
        });
    }
    let status = if build.status == "success"
        && test.status == "success"
        && !findings.iter().any(|finding| finding.severity == "error")
    {
        "success"
    } else {
        "failed"
    }
    .to_string();

    Ok(AnalyzeData {
        project_root: resolved.project_root.display().to_string(),
        status,
        build_status: build.status,
        test_status: test.status,
        diagnostics: build.diagnostics,
        findings,
        test_stdout: test.stdout,
        test_stderr: test.stderr,
    })
}

#[derive(Debug)]
struct TestRun {
    status: String,
    stdout: String,
    stderr: String,
}

fn forge_test(project_root: &std::path::Path) -> AppResult<TestRun> {
    let output = Command::new("forge")
        .arg("test")
        .arg("--root")
        .arg(project_root)
        .arg("--color")
        .arg("never")
        .output()
        .map_err(|err| {
            AppError::user(
                "analysis_test_failed",
                format!("Failed to run forge test: {err}"),
                Some("Install Foundry and make sure `forge` is on PATH.".to_string()),
            )
        })?;

    Ok(TestRun {
        status: if output.status.success() {
            "success".to_string()
        } else {
            "failed".to_string()
        },
        stdout: String::from_utf8_lossy(&output.stdout).to_string(),
        stderr: String::from_utf8_lossy(&output.stderr).to_string(),
    })
}

fn finding_from_diagnostic(diagnostic: &build::Diagnostic) -> AnalyzeFinding {
    AnalyzeFinding {
        severity: diagnostic.severity.clone(),
        source: diagnostic.source.clone(),
        message: diagnostic.message.clone(),
        file: diagnostic.file.clone(),
        line: diagnostic.line,
        column: diagnostic.column,
    }
}

fn print_human_findings(data: &AnalyzeData) {
    println!("Analysis failed: {}", data.project_root);
    println!("  build: {}", data.build_status);
    println!("  test: {}", data.test_status);
    for finding in &data.findings {
        let location = match (&finding.file, finding.line, finding.column) {
            (Some(file), Some(line), Some(column)) => format!("{file}:{line}:{column}"),
            (Some(file), _, _) => file.clone(),
            _ => finding.source.clone(),
        };
        println!("  {} {} {}", finding.severity, location, finding.message);
    }
}

File: packages/cli/src/commands/analyze.ts
import { resolveTarget } from "@consol/core";
import { runForgeBuild, runForgeTest } from "@consol/foundry";
import { createSuccessEnvelope } from "@consol/protocol";
import type { GlobalArgs } from "../args";
import type { CliEnv, CliResult } from "../main";
import { VERSION } from "../version";
import { parseBuildDiagnostics, type BuildDiagnostic } from "./diagnostics";

export type AnalyzeFinding = {
  readonly severity: "error" | "warning";
  readonly source: string;
  readonly message: string;
  readonly file: string | null;
  readonly line: number | null;
  readonly column: number | null;
};

export type AnalyzeData = {
  readonly project_root: string;
  readonly status: "success" | "failed";
  readonly build_status: "success" | "failed";
  readonly test_status: "success" | "failed";
  readonly diagnostics: readonly BuildDiagnostic[];
  readonly findings: readonly AnalyzeFinding[];
  readonly test_stdout: string;
  readonly test_stderr: string;
};

export type RunAnalyzeCommandInput = {
  readonly globals: GlobalArgs;
  readonly commandArgs: readonly string[];
  readonly cwd: string;
  readonly env: CliEnv;
};

export async function runAnalyzeCommand(input: RunAnalyzeCommandInput): Promise<CliResult> {
  const resolved = resolveTarget({
    cwd: input.cwd,
    ...(input.globals.project === undefined ? {} : { projectRoot: input.globals.project }),
  });
  const build = await runForgeBuild({
    cwd: resolved.projectRoot,
    projectRoot: resolved.projectRoot,
    env: input.env,
  });
  const test = await runForgeTest({
    cwd: resolved.projectRoot,
    projectRoot: resolved.projectRoot,
    env: input.env,
  });
  const diagnostics = parseBuildDiagnostics(build.stdout, build.stderr);
  const findings = diagnostics.map(findingFromDiagnostic);
  if (!test.ok) {
    findings.push({
      severity: "error",
      source: "forge test",
      message: "Foundry tests failed.",
      file: null,
      line: null,
      column: null,
    });
  }

  const data: AnalyzeData = {
    project_root: resolved.projectRoot,
    status: build.ok && test.ok && !findings.some((finding) => finding.severity === "error") ? "success" : "failed",
    build_status: build.ok ? "success" : "failed",
    test_status: test.ok ? "success" : "failed",
    diagnostics,
    findings,
    test_stdout: test.stdout,
    test_stderr: test.stderr,
  };

  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: "analyze",
        project_root: data.project_root,
      },
    });
    return { exitCode: data.status === "success" ? 0 : 1, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }

  if (data.status === "success") {
    return { exitCode: 0, stdout: `Analysis passed: ${data.project_root}\n`, stderr: "" };
  }

  return {
    exitCode: 1,
    stdout: formatHumanFindings(data),
    stderr: "ConSol analysis found issues.\n",
  };
}

function findingFromDiagnostic(diagnostic: BuildDiagnostic): AnalyzeFinding {
  return {
    severity: diagnostic.severity,
    source: diagnostic.source,
    message: diagnostic.message,
    file: diagnostic.file,
    line: diagnostic.line,
    column: diagnostic.column,
  };
}

function formatHumanFindings(data: AnalyzeData): string {
  const lines = [`Analysis failed: ${data.project_root}`, `  build: ${data.build_status}`, `  test: ${data.test_status}`];
  for (const finding of data.findings) {
    lines.push(`  ${finding.severity} ${findingLocation(finding)} ${finding.message}`);
  }
  return `${lines.join("\n")}\n`;
}

function findingLocation(finding: AnalyzeFinding): string {
  if (finding.file !== null && finding.line !== null && finding.column !== null) {
    return `${finding.file}:${finding.line}:${finding.column}`;
  }
  return finding.file ?? finding.source;
}
